### PR TITLE
Adding tests for rect extractions

### PR DESCRIPTION
**Pull request by kreuzberger**
Added tests for rect extraction for sphinx-simplepdf / weasyprint generated pdf.
Tests checks for textbox extraction from codeblocks, admonitions and tables.

The tests for table did not work as expected. Instead of extracting colored table cells as rect, the 3 table row shown with alternating colors is extracted as whole.
Attached is a picture from visual debug.

The tests now works, asuming the "wrong" number of rects (i would expect 7). See attached file.
All other extractions work like expected.
libpdf_rect_from_table

### CODE REVIEW

**Comment by kreuzberger** [comment]
The failing test has nothing to do with the test implemenation, the test is missing a required executable!
This now explains why i had to patch /etc/ImageMagic Policies on my ubuntu machine.!

This seems to be a feature of the visual debug. A hint in the doc would help.

File: tests/conftest.py [hunk 1]
```diff
@@ -31,6 +31,11 @@
 # test PDFs from official python documentation
 PDF_PYTHON_LOGGING = os.path.join(os.path.dirname(__file__), "pdf", "howto-logging.pdf")
 
+# test PDF for rect extraction generateby by sphinx-simplepdf
```
Comment (ubmarco) [inline tests/conftest.py:33]:
```suggestion
# test PDF for rect extraction generated by sphinx-simplepdf
```

**Comment by ubmarco** [comment]
Thanks a lot for your PR, I really appreciate new tests for the library.

I cannot push to your branch as the fork is created on your organization, not on your personal account.
So I added a commit to your branch and created a new PR from it to see the changes in CI https://github.com/useblocks/libpdf/pull/37.
The ruff linting is non-voting for now, but I want to enable it over time for more and more files. Once a file is touched I will add it to the files in the tox.ini `lint` environment.

File: tests/test_rects.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    search for given chapter in the objects.
+
+    :return: found chapter
+    """
+    chapters = objects.flattened.chapters
+    ret_chapter = None
+    assert len(chapters) > 0
+    for chapter in chapters:
+        if chapter.title == chapter_name:
+            ret_chapter = chapter
```
Comment (ubmarco) [inline tests/test_rects.py]: add a break to the loop?

File: tests/test_rects.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    smart_page_crop = (
+        True  # remove header and footers so rects IN chapters are left only.
+    )
+
+    objects = libpdf.load(
+        PDF_RECTS_EXTRACTION,
+        smart_page_crop=smart_page_crop,
+        visual_debug=False,
+        visual_debug_output_dir=tmpdir.join("visual_debug_dir"),
+        visual_split_elements=True,
```
Comment (ubmarco) [inline tests/test_rects.py]: these 2 should not be needed if you set `visual_debug=False`

**Review by ubmarco — CHANGES_REQUESTED** [review]

**Comment by ubmarco** [comment]
PR https://github.com/useblocks/libpdf/pull/37 fails as expected. I propose to cherry-pick my commits into your branch and fix the mentioned issues from my review.

**Comment by ubmarco** [comment]
For the rect count in your PDF, this is how the PDF is made, the header row actually has 3 rectangles while the row is just one. If you zoom in extremely, you also see it, e.g. here in Firefox's pdf.js based reader:
image
If you look closely, you see bright vertical lines in the header, but not in the row.

File: tests/test_rects.py (hunk 3 above)
Comment (kreuzberger) [reply tests/test_rects.py]: Yes, but just want to change visual_debug to True without chanching the later ones.
Could be set by using a variable?

vs_debug = False

visual_debug = vs_debug
visual_split_elements = not vs_debug
visual_debug_output = tmpdir.join("visual_debug_dir") if vs_debug else ""

Or just leave it :smile:

File: tests/test_rects.py (hunk 2 above)
Comment (kreuzberger) [reply tests/test_rects.py]: using list comprehension with next and default to None

File: tests/conftest.py (hunk 1 above)
Comment (ubmarco) [reply tests/conftest.py:33]: Not urgent, but why not fix it when spotted in a review

File: tests/test_rects.py (hunk 3 above)
Comment (ubmarco) [reply tests/test_rects.py]: Still don't understand this. Why would you provide fields that have no meaning in this context?
image
The libpdf run will not populate this directory nor split elements, so why passing the params. Maybe I misunderstand your use case here. :thinking:

**Review by ubmarco — CHANGES_REQUESTED** [review]
almost there

**Comment by kreuzberger** [comment]
> almost there
If i set visual_debug to True, the debugging is configured. This is my main intention.

**Review by ubmarco — APPROVED** [review]

**Pull request merged**